packages/fxa-auth-server/lib/routes/oauth.js

```diff
@@ -215,6 +219,56 @@ module.exports = (log, config, oauthdb, db, mailer, devices) => {
         return grant;
       },
     },
+    {
+      method: 'POST',
+      path: '/oauth/destroy',
+      options: {
+        validate: {
+          headers: clientAuthValidators.headers,
+          payload: {
+            client_id: clientAuthValidators.clientId,
+            client_secret: clientAuthValidators.clientSecret,
+            token: Joi.alternatives().try(
+              validators.accessToken,
+              validators.refreshToken
+            ),
+            // The spec says we have to ignore invalid token_type_hint values,
+            // but no way I'm going to accept an arbitrarily-long string here...
+            token_type_hint: Joi.string()
+              .max(64)
+              .optional(),
+          },
+        },
+        response: {},
+      },
+      handler: async function(request) {
+        // This endpoint implements the API for token revocation from RFC7009,
+        // which says that if we can't find the token using the provided token_type_hint
+        // then we MUST search other possible types of token as well. So really
+        // token_type_hint just tells us what *order* to try different token types in.
+        let methodsToTry = ['revokeAccessToken', 'revokeRefreshToken'];
+        if (request.payload.token_type_hint === 'refresh_token') {
```

Comment:
If the passed token is a JWT, does it make sense to just skip checking for a refreshToken? Of course that makes the assumption that refresh tokens will never be JWTs...

Reply:
I think this is reasonable, and we can make it more generic by checking against the validators rather than hard-coding a check for whether it's a JWT.